docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+chosen to replace the old one, should pay a *time window* necessary to "transfer"
+the ownership over the shared `PV`.
+
+This means that, unlike the *Active-Active* solution, it is not guaranteed that
+the subsequent access made by a client after a failed one is always successful.
+
+It can happen that a client issues an access to the S3 service while the system is
+translating the passive pipeline into the active state; in this case, the issued
+access is expected to fail.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw)
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+```
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw)
+    ingress ==> radosgw2(radosgw) == ownership transfer ==> PV((PV))
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: white
+    linkStyle 2 stroke: red
+```
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw)
+    ingress ==> radosgw2(radosgw) ==> PV((PV))
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: white
+    linkStyle 2 stroke: green
+    linkStyle 3 stroke: green
+```
+
+## One Active pipeline
+
+This model is logically equivalent to the *Active-Passive* model with the difference
+that only one pipeline is allocated on the cluster at a time.
+
+In the *One pipeline* model, the only notable perceived difference between
+the *Active-Passive* model are the worst performances obtained in the event of failure.
+
+In this scenario, the system should pay also the time needed to fully load a new
+pipeline, that unlike that *Active-Passive* model, is not preloaded.
+
+So, supposing that the new pipeline has scheduled to load on a node where the
+s3gw image is not cached, the system should pay the time needed to download
+the image from the registry before starting it.
+
+## Investigation's Rationale
+
+The 3 models described above have different performances and different implementation
+efforts. While the *Active-Active* model is expected to require a significant
+development effort due to its inherent complex nature, the *Passive-Active* model
+seems to offer a lower degree of implementation difficulty.
+
+In particular, the *One Active pipeline* model, is expected to work with nothing
+but the primitives normally available on any Kubernetes cluster.
+
+Given that the backend s3gw architecture is composed by:
+
+- one *stateless* `radosgw` process associated with a *stateful* `PV`
+
+it is supposed that, on `radosgw` process failure, a simple reschedule
+of its POD is enough to fulfill this simple HA model.
+A clarification is needed here: the `PV` *state* is guaranteed by a third party
+service, that is: **Longhorn**.
+
+Longhorn ensures that a `PV` (along with its content) is always available on
```

Comment:
While we consider failure modes, this is only true if the cluster still has quorum and there may be LH options that affect availability and recovery. It's also important to call out that obviously we can't achieve higher availability or reliability than the underlying LH volumes.

Reply:
What do you mean exactly with  _if the cluster still has quorum_ ?

> It's also important to call out that obviously we can't achieve higher availability or reliability than the underlying LH volumes.

I'd say that the s3gw has more to do with the concept of S3 _availability_ while LH has the responsibility to maintain the state of the underlying data in case of severe failure.
We could even say that our model  prefers data preservation over the (S3) data _availability_ and thus that in certain cases it is preferable to completely stop the  _availability_ to  maximize the current  data state.
That is, we could prefer the lesser bad (stop the access to S3 clients)  to safeguard the data state.
(I'm just thinking aloud here)